### ISSUE TITLE
SharedMemoryArray is too slow

### ISSUE THREAD

**Issue by tridemax** [google/grain#851]
I've extensively tested the MP implementation in grain and it's very slow. 

![Image](https://github.com/user-attachments/assets/cf75646a-774c-4b26-b775-24c7a44830a1)

The $batch in the middle (violet) is the stacking function that takes 21ms to stack batch of 128. So I expect it should be like 25-40ms as a whole but the 'next' call takes whooping 224 ms. On this profile trace above you can see the awaiting, and then deleting the SharedMemoryArrays, dominates the 'next' call.

I can see how awaiting can struggle because workers are not fast enough (and this is not the case) but I can't see why simple deleting operation could take 1+ms for each array. I believe it just kills the purpose of the shared memory array and naive push through 'multiprocessing.Queue' would be much faster.

Any plans on improving the MP implementation?

**Comment by tridemax** [comment]
For those who struggle with very slow MP loading here is some short explanation how to go from above 224 ms to 0.031 ms (yeah, microseconds):
![Image](https://github.com/user-attachments/assets/969dcde2-507d-4713-9195-5ed0472d9364)

We still rely on Grain's deterministic pipeline and have all the good things from state serialization, etc.

Create a multi-process pool executor:
```
pool_executor = None
pool_executor_num_workers = 0

def _dummy_call(idx):
    time.sleep(0.1)
    return idx


def create_pool_executor(max_workers=cpu_count()):
    global pool_executor
    global pool_executor_num_workers

    pool_executor_num_workers = max_workers

    pool_executor = concurrent.futures.ProcessPoolExecutor(max_workers=max_workers)
    pool_executor._call_queue = multiprocessing.Queue()  # create unbound queue to not to hinder pushing performance
    pool_executor._call_queue._ignore_epipe = True

    # Create processes
    _futures = [pool_executor.submit(_dummy_call, i) for i in range(max_workers)]
    return pool_executor
```

Main idea is that your mapping function just submits the path to the record to multi-process worker and pushes future for this operation down the pipeline:
```
def _map_fn(self, path: str, rnd: np.random.Generator, fn) -> Future:
    return self._pool_executor.submit(fn, path, rnd)
```

Your batching function just requests the results:
```
def _batch_fn(self, batch: Sequence[Future]) -> 'YourReturnType':
    # Get all the results
    results = [future.result() for future in batch]
    ...
    return batch
```

Your `@staticmethod` loading function can return dict/tuple of numpy arrays, that works perfectly fast. Static method is required, otherwise the whole class would be serialized to the worker on each call:
```
@staticmethod
def load_record(path: str, ...) -> 'BatchEntry':
    ...
    return BatchEntry(np_array1=arr1, np_array2=arr2, path=path)
```

Then you define your dataset:
```
        train_dataset = train_dataset.random_map(  # random_map here for rnd parameter, could be just map
            transform=partial(self._map_fn,
                              fn=partial(self.load_record,  # Make sure the loading function is @staticmethod
                                         ...  # Your additional function parameters goes here if needed, otherwise skip partial()
                                        ),
            ),
            seed=self.seed)
        train_dataset = train_dataset.batch(self.batch_size, batch_fn=self._batch_fn)
        train_dataset = train_dataset.to_iter_dataset()
        train_dataset = grain.experimental.ThreadPrefetchIterDataset(train_dataset, prefetch_buffer_size=self.prefetch_buffer_size)
```

The last line uses `ThreadPrefetchIterDataset` that prefetches few batches **in other thread**, so while Jax is chewing the current batch and releases the GIL - your another thread prepares the next batch(es) for you. Small value of `prefetch_buffer_size==2` usually works for me.

Then you use your batch normally:
```
batch = next(iter_train)
batch = OnDeviceBatch(batch)  # this converts from numpy to jax.Array and optionally converts (like to bfloat16)
result = jitted_loss_func(batch)  # call your model
```

**Comment by tridemax** [comment]
One more note: I've tried `ThreadPrefetchIterDataset` just on top of `dataset = dataset.mp_prefetch()` in my original comment. It didn't work because the ~220ms on awaiting/cleanup is still there and my step time is half of that.

So I'm still not sure about design decision of introducing SharedMemoryArray instead of simple queue serialization/deserialization of np.ndarrays.

And one more thing. One would need to play with `ulimit -n` on their system if MP prefetch is going over separate batch entries, as you can easily hit the limit on open files with shm, especially if your batch entries has multiple arrays (more than 2). One can argue he could use MP prefetch over the batches but that would consume a lot of memory with big enough number of workers, while single worker would still struggle to load and process the entire batch fast enough.

**Comment by iindyk** [comment]
`mp_prefetch` is normally done after batching, so that larger arrays are passed over shared memory. If you want to switch to serialization set the [shm array size threshold](https://google-grain.readthedocs.io/en/latest/_autosummary/grain.experimental.DatasetOptions.html#grain.experimental.DatasetOptions.min_shm_size) to a higher value -- every array below that size will be serialized.

we ought to tune the default for the option to something reasonable.
